llvm/Analysis/AliasAnalysis.h does not build in C++23 mode

Repro: (clang / llvm 18.1.8 as provided by Arch Linux)

foo.cpp:
```cpp
#include <llvm/LinkAllPasses.h>
```

with libstdc++-14:
```bash
$ clang++ -std=c++20 -c foo.cpp
(ok)

$ g++ -std=c++20 -c foo.cpp
(ok)

$ g++ -std=c++23 -c foo.cpp
(ok)

$ clang++ -std=c++23 -c foo.cpp
In file included from foo.cpp:2:
In file included from /usr/include/llvm/LinkAllPasses.h:17:
In file included from /usr/include/llvm/ADT/Statistic.h:32:
In file included from /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/memory:78:
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/unique_ptr.h:91:16: error: invalid application of 'sizeof' to an incomplete type 'llvm::AAResults::Concept'
   91 |         static_assert(sizeof(_Tp)>0,
      |                       ^~~~~~~~~~~
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/unique_ptr.h:398:4: note: in instantiation of member function 'std::default_delete<llvm::AAResults::Concept>::operator()' requested here
  398 |           get_deleter()(std::move(__ptr));
      |           ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/stl_construct.h:88:15: note: in instantiation of member function 'std::unique_ptr<llvm::AAResults::Concept>::~unique_ptr' requested here
   88 |         __location->~_Tp();
      |                      ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/stl_construct.h:149:12: note: in instantiation of function template specialization 'std::destroy_at<std::unique_ptr<llvm::AAResults::Concept>>' requested here
  149 |       std::destroy_at(__pointer);
      |            ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/stl_construct.h:163:11: note: in instantiation of function template specialization 'std::_Destroy<std::unique_ptr<llvm::AAResults::Concept>>' requested here
  163 |             std::_Destroy(std::__addressof(*__first));
      |                  ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/stl_construct.h:193:35: note: in instantiation of function template specialization 'std::_Destroy_aux<false>::__destroy<std::unique_ptr<llvm::AAResults::Concept> *>' requested here
  193 |         return std::_Destroy_aux<false>::__destroy(__first, __last);
      |                                          ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/alloc_traits.h:944:12: note: in instantiation of function template specialization 'std::_Destroy<std::unique_ptr<llvm::AAResults::Concept> *>' requested here
  944 |       std::_Destroy(__first, __last);
      |            ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/stl_vector.h:735:7: note: in instantiation of function template specialization 'std::_Destroy<std::unique_ptr<llvm::AAResults::Concept> *, std::unique_ptr<llvm::AAResults::Concept>>' requested here
  735 |         std::_Destroy(this->_M_impl._M_start, this->_M_impl._M_finish,
      |              ^
/usr/include/llvm/Analysis/AliasAnalysis.h:311:3: note: in instantiation of member function 'std::vector<std::unique_ptr<llvm::AAResults::Concept>>::~vector' requested here
  311 |   AAResults(const TargetLibraryInfo &TLI) : TLI(TLI) {}
      |   ^
/usr/include/llvm/Analysis/AliasAnalysis.h:604:9: note: forward declaration of 'llvm::AAResults::Concept'
  604 |   class Concept;
      |         ^
1 error generated.
```

with libc++-18:
```bash
$ clang++ -std=c++23 -c foo.cpp -stdlib=libc++
In file included from foo.cpp:2:
In file included from /usr/include/llvm/LinkAllPasses.h:17:
In file included from /usr/include/llvm/ADT/Statistic.h:32:
In file included from /usr/bin/../include/c++/v1/memory:939:
In file included from /usr/bin/../include/c++/v1/__memory/shared_ptr.h:33:
/usr/bin/../include/c++/v1/__memory/unique_ptr.h:64:19: error: invalid application of 'sizeof' to an incomplete type 'llvm::AAResults::Concept'
   64 |     static_assert(sizeof(_Tp) >= 0, "cannot delete an incomplete type");
      |                   ^~~~~~~~~~~
/usr/bin/../include/c++/v1/__memory/unique_ptr.h:266:7: note: in instantiation of member function 'std::default_delete<llvm::AAResults::Concept>::operator()' requested here
  266 |       __ptr_.second()(__tmp);
      |       ^
/usr/bin/../include/c++/v1/__memory/unique_ptr.h:236:71: note: in instantiation of member function 'std::unique_ptr<llvm::AAResults::Concept>::reset' requested here
  236 |   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 ~unique_ptr() { reset(); }
      |                                                                       ^
/usr/bin/../include/c++/v1/__memory/construct_at.h:67:11: note: in instantiation of member function 'std::unique_ptr<llvm::AAResults::Concept>::~unique_ptr' requested here
   67 |   __loc->~_Tp();
      |           ^
/usr/bin/../include/c++/v1/__memory/allocator_traits.h:316:10: note: in instantiation of function template specialization 'std::__destroy_at<std::unique_ptr<llvm::AAResults::Concept>, 0>' requested here
  316 |     std::__destroy_at(__p);
      |          ^
/usr/bin/../include/c++/v1/vector:926:23: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<std::unique_ptr<llvm::AAResults::Concept>>>::destroy<std::unique_ptr<llvm::AAResults::Concept>, void, void>' requested here
  926 |       __alloc_traits::destroy(__alloc(), std::__to_address(--__soon_to_be_end));
      |                       ^
/usr/bin/../include/c++/v1/vector:920:5: note: in instantiation of member function 'std::vector<std::unique_ptr<llvm::AAResults::Concept>>::__base_destruct_at_end' requested here
  920 |     __base_destruct_at_end(this->__begin_);
      |     ^
/usr/bin/../include/c++/v1/vector:490:16: note: in instantiation of member function 'std::vector<std::unique_ptr<llvm::AAResults::Concept>>::__clear' requested here
  490 |         __vec_.__clear();
      |                ^
/usr/bin/../include/c++/v1/vector:501:67: note: in instantiation of member function 'std::vector<std::unique_ptr<llvm::AAResults::Concept>>::__destroy_vector::operator()' requested here
  501 |   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~vector() { __destroy_vector (*this)(); }
      |                                                                   ^
/usr/include/llvm/Analysis/AliasAnalysis.h:311:3: note: in instantiation of member function 'std::vector<std::unique_ptr<llvm::AAResults::Concept>>::~vector' requested here
  311 |   AAResults(const TargetLibraryInfo &TLI) : TLI(TLI) {}
      |   ^
/usr/include/llvm/Analysis/AliasAnalysis.h:604:9: note: forward declaration of 'llvm::AAResults::Concept'
  604 |   class Concept;
      |         ^
1 error generated.
```

--- a/llvm/include/llvm/Analysis/AliasAnalysis.h
+++ b/llvm/include/llvm/Analysis/AliasAnalysis.h
@@ -315,7 +315,7 @@ class AAResults {
 public:
   // Make these results default constructable and movable. We have to spell
   // these out because MSVC won't synthesize them.
-  AAResults(const TargetLibraryInfo &TLI) : TLI(TLI) {}
+  AAResults(const TargetLibraryInfo &TLI);
   AAResults(AAResults &&Arg);
   ~AAResults();
 

--- a/llvm/lib/Analysis/AliasAnalysis.cpp
+++ b/llvm/lib/Analysis/AliasAnalysis.cpp
@@ -73,6 +73,8 @@ static cl::opt<bool> EnableAATrace("aa-trace", cl::Hidden, cl::init(false));
 static const bool EnableAATrace = false;
 #endif
 
+AAResults::AAResults(const TargetLibraryInfo &TLI) : TLI(TLI) {}
+
 AAResults::AAResults(AAResults &&Arg)
     : TLI(Arg.TLI), AAs(std::move(Arg.AAs)), AADeps(std::move(Arg.AADeps)) {}
 

